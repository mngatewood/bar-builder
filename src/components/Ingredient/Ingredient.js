import React, { Component } from 'react';
import './Ingredient.css';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { addIngredient, removeIngredient } from '../../actions';

export class Ingredient extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: this.props.inInventory
    };
  }

  handleChange = () => {
    const value = this.props.value;
    this.state.selected ?
      this.props.removeIngredient(value) :
      this.props.addIngredient(value);
    this.setState({ selected: !this.state.selected });
  };

  render() {
    return <li
      className="ingredient">
      <input
        type="checkbox"
        value={this.props.value}
        checked={this.state.selected}
        onChange={this.handleChange} />
      <span>
        &nbsp;{this.props.value}
      </span>
    </li>;
  }

}

export const mapDispatchToProps = (dispatch) => ({
  addIngredient: ingredient => dispatch(addIngredient(ingredient)), 
  removeIngredient: ingredient => dispatch(removeIngredient(ingredient)) 
});

Ingredient.propTypes = {
  addIngredient: PropTypes.func,
  removeIngredient: PropTypes.func,
  ingredient: PropTypes.object,
  value: PropTypes.string,
  inInventory: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool
  ])
};

export default connect(null, mapDispatchToProps)(Ingredient);
