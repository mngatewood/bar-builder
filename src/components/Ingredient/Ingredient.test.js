import React from 'react';
import { shallow } from 'enzyme';
import { Ingredient } from './Ingredient';

describe("Ingredient", () => {

  let addIngredient, removeIngredient, wrapper;

  beforeEach(() => {

    addIngredient = jest.fn();
    removeIngredient = jest.fn();
    wrapper = shallow(
      <Ingredient
        inInventory="false"
        addIngredient={addIngredient}
        removeIngredient={removeIngredient}
      />);

  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls addIngredient if it is selected', () => {
    wrapper.instance().handleChange();
    expect(addIngredient).toHaveBeenCalled;
  });

  it('calls removeIngredient if it is deselected', () => {
    wrapper = shallow(
      <Ingredient
        inInventory="true"
        addIngredient={addIngredient}
        removeIngredient={removeIngredient}
      />);

    wrapper.instance().handleChange();
    expect(removeIngredient).toHaveBeenCalled;
  });

});