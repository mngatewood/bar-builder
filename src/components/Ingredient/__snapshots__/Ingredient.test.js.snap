// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredient matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Ingredient
    addIngredient={[Function]}
    inInventory="false"
    removeIngredient={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked="false"
          onChange={[Function]}
          type="checkbox"
          value={undefined}
/>,
        <span>
           
</span>,
      ],
      "className": "ingredient",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": "false",
          "onChange": [Function],
          "type": "checkbox",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            " ",
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          " ",
          undefined,
        ],
        "type": "span",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked="false"
            onChange={[Function]}
            type="checkbox"
            value={undefined}
/>,
          <span>
             
</span>,
        ],
        "className": "ingredient",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": "false",
            "onChange": [Function],
            "type": "checkbox",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              " ",
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            " ",
            undefined,
          ],
          "type": "span",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
